lib/Sema/CSApply.cpp

```diff
@@ -3320,14 +3320,21 @@ namespace {
       auto destValueType
         = finalResultType->lookThroughAllAnyOptionalTypes(destOptionals);
 
-      // When performing a bridging operation, if the destination value type
-      // is 'AnyObject', leave any extra optionals on the source in place.
-      if (castKind == OptionalBindingsCastKind::Bridged &&
-          srcOptionals.size() > destOptionals.size() &&
-          destValueType->isAnyObject()) {
-        srcType = srcOptionals[destOptionals.size()];
-        srcOptionals.erase(srcOptionals.begin() + destOptionals.size(),
-                           srcOptionals.end());
+      auto isBridgeToAnyObject =
+        castKind == OptionalBindingsCastKind::Bridged &&
+        destValueType->isAnyObject();
+
+      // If the destination value type is 'AnyObject' when performing a
+      // bridging operation, or if the destination value type is an
+      // archetype (and could therefore be an Optional type at runtime),
+      // leave any extra optionals on the source in place.
+      if (isBridgeToAnyObject || destValueType->is<ArchetypeType>()) {
```

Comment:
You should disallow class-archetypes.  We have a function in DynamicCasts.cpp, canDynamicallyBeOptionalType().  It's not quite what you want because it allows existentials, but please extract something in common that can be used by both places.  Having a parameter indicating whether we're asking whether the type could *be* an optional type (what you're asking here) or *store* an optional type (what DynamicCasts.cpp is asking) would be sensible.

Reply:
Very helpful, thanks @rjmccall! I've gone ahead and made those changes, let me know if they're what you had in mind.